dominion/src/dominion/arena.rs

```diff
@@ -1,40 +1,515 @@
+                _ => Err(Error::InvalidCardLocation),
+            },
+            TurnPhase::Buy {
+                remaining_buys,
+                total_wealth,
+            } => match location {
+                Location::Hand => match self.players[player_id].hand.remove_item(&card) {
+                    Some(card) => match card.treasure() {
+                        Some(i) => {
+                            *total_wealth += i;
+                            self.players[player_id].played.push(card);
+                            Ok(())
+                        }
+                        None => {
+                            self.players[player_id].hand.push(card);
+                            Err(Error::InvalidCardChoice)
+                        }
+                    },
+                    None => Err(Error::InvalidCardChoice),
+                },
+                Location::Supply => match self.supply.get_mut(card) {
+                    Some(supply_count) => {
+                        if *supply_count == 0 {
+                            Err(Error::NoMoreCards)
+                        } else if *remaining_buys == 0 {
+                            Err(Error::NoMoreBuys)
+                        } else if *total_wealth < card.cost() {
+                            Err(Error::NotEnoughWealth)
+                        } else {
+                            *supply_count -= 1;
+                            *remaining_buys -= 1;
+                            *total_wealth -= card.cost();
+                            self.players[player_id].discard_pile.push(card);
+                            Ok(())
+                        }
+                    }
+                    None => Err(Error::InvalidCardChoice),
+                },
+                _ => Err(Error::InvalidCardLocation),
+            },
+        }
+    }
+
+    pub fn check_hand(&self, player_id: usize) -> Result<&CardVec> {
+        Ok(&self.get_player(player_id)?.hand)
     }
 
-    pub fn players(&mut self) -> &mut Vec<Player> {
-        &mut self.players
+    pub fn check_discard_pile(&self, player_id: usize) -> Result<&CardVec> {
+        Ok(&self.get_player(player_id)?.discard_pile)
+    }
+
+    pub fn check_played(&self, player_id: usize) -> Result<&CardVec> {
+        Ok(&self.get_player(player_id)?.played)
+    }
+
+    pub fn check_deck(&self, player_id: usize, card: CardKind) -> Result<bool> {
+        Ok(self.get_player(player_id)?.in_deck(card))
+    }
+
+    fn check_active(&mut self, player_id: usize) -> Result<()> {
+        if player_id >= self.players.len() {
+            Err(Error::InvalidPlayerId)
+        } else if player_id != self.turn.player_id {
+            Err(Error::InactivePlayer)
+        } else {
+            Ok(())
+        }
     }
 
     fn start_game(&mut self) {
-        for p in self.players() {
+        for p in self.players.iter_mut() {
             p.cleanup();
         }
     }
+
+    fn get_player(&self, player_id: usize) -> Result<&Player> {
```

Comment:
`s/get_player/player/g`
consider adding
`current_player` and `current_player_mut` which use turn.player_id

Reply:
Done and done. However, I'm not sure how we will use `current_player` and `current_player_mut` right now.